source/common/common/radix_tree.h

```diff
@@ -0,0 +1,324 @@
+#pragma once
+
+#include <algorithm>
+#include <vector>
+
+#include "source/common/common/assert.h"
+
+#include "absl/container/flat_hash_map.h"
+#include "absl/strings/string_view.h"
+#include "absl/types/optional.h"
+
+namespace Envoy {
+template <class Value> class RadixTree {
+  static constexpr int32_t NoNode = -1;
+  struct RadixTreeNode {
+    std::string prefix_;
+    Value value_{};
+    // Hash map for O(1) child lookup by first character
+    absl::flat_hash_map<uint8_t, RadixTreeNode> children_;
+
+    /**
+     * Insert a key-value pair into this node
+     * @param key the full key being inserted
+     * @param search the remaining search key
+     * @param value the value to insert
+     */
+    void insert(absl::string_view key, absl::string_view search, Value value) {
```

Comment:
I think `key` isn't used at all in this function - it's just passed down recursively and ignored at the bottom.